[C+B] Add a world broadcast and broadcastMessage function

kezz opened the pull request:
This PR adds two new functions to the CraftWorld class, broadcast and broadcastMessage. They exactly mimic CraftServer's broadcast and broadcastMessage except CraftWorld's version will only broadcast the message to players in that world.

I added it because I use lots of plugins that iterate through every single player, check the world they are in and then send them a message. It is a lot easier and less time consuming to use something like `player.getWorld().broadcastMessage("Hello people in my world!");`

CraftBukkit and Bukkit both compile successfully using Maven with this awesome addition.

To demonstrate this I have made a plugin that just has one command

> /test [worldName] [permission]  

If you want to see this in action for yourself you can download the plugin [here](https://www.dropbox.com/s/mgwgxzw6r44b4xv/WBT.jar?dl=1) and the modified CraftBukkit [here](https://www.dropbox.com/s/snb87sdiqxydp73/craftbukkit-1.4.7-R1.1-SNAPSHOT.jar?dl=1).

See https://github.com/Bukkit/CraftBukkit/pull/1033
See https://bukkit.atlassian.net/browse/BUKKIT-3638

ScoreUnder committed:
Move world generator warning to CraftBukkit. Fixes BUKKIT-2565

oloflarsson committed:
Add ability to modify ThrownPotion properties. Adds BUKKIT-3197

psycowithespn committed:
Add ability to change player item in hand. Adds BUKKIT-3318

Peter200lx committed:
Add inverted flag support to TrapDoor. Fixes BUKKIT-3390

nitnelave committed:
Added the hasGravity method to Blocks. Adds BUKKIT-3832

crast committed:
Refactor common metadata code into base class. Fixes BUKKIT-3624

Implementing the MetadataValue interface is significant work due to
having to provide a large amount of conversion stub methods. This commit
adds a new optional abstract base class to aid in implementation.

crast committed:
Improve speed and memory use of FixedMetadataValue. Fixes BUKKIT-1460

FixedMetadataValue currently just extends LazyMetadataValue with a value
that never changes. While this works it is a lot of unneeded overhead
that causes FixedMetadataValue to be a lot slower and use a lot more
memory than one would expect. To correct this we store the value directly
in FixedMetadataValue and override the the appropriate methods to use it.

Ideally we would modify FixedMetadataValue to no longer extend
LazyMetadataValue as this would give a very large memory savings. However,
this is not currently done for backwards compatibility reasons.

Commit:
Added broadcast and broadcastMessage

Commit:
Added broadcast and broadcastMessage

andrepl committed:
Add Fish (Hook) entity to PlayerFishEvent. Adds BUKKIT-1025

turt2live commented:
There should be a leaky ticket and a use case applied with this

riking commented on src/main/java/org/bukkit/World.java:
```diff
@@ -1061,6 +1061,23 @@
+    public int broadcast(String message, String permission);
+    
+    /**
+     * Broadcasts the specified message to every player in this world
```
Suggestion:

```
Broadcasts the specified message to every player in this world with the {@link org.bukkit.Server#BROADCAST_CHANNEL_USERS} permission
```

kezz replied:
Hmm could do... My only thought is that it it not specified on the normal broadcast... Hmm...

kezz commented:
@turt2live All the information you specified is available on the matching CraftBukkit pull request which I linked to above.

riking commented:
@kezz101 You generally put all the information on both.
Also, try adding a [C+B] to the issue title

kezz commented:
@riking All done :+1:

natemort commented on src/main/java/org/bukkit/World.java:
```diff
@@ -1061,6 +1061,23 @@
      * @return True if rule exists
      */
     public boolean isGameRule(String rule);
+    
```
You've got spaces on this line.

natemort commented on src/main/java/org/bukkit/World.java:
```diff
@@ -1061,6 +1061,23 @@
+     * @return the number of players who recieved the message
+     */
+    public int broadcast(String message, String permission);
+    
```
Spaces.

natemort committed:
BlockState for Command Blocks. Adds BUKKIT-3805.

SagaciousZed committed:
Pulling all pending Bukkit-JavaDoc changes

gmcferrin committed:
Add methods to check for conflicting enchantments. Adds BUKKIT-3830

andrepl committed:
Allow fishing success rate to be adjustable. Adds BUKKIT-3837

mbax committed:
Add Scoreboard API and Command. Adds BUKKIT-3776, BUKKIT-3834

The implementation is designed around having both a main scoreboard and
numberous plugin managed scoreboards that can be displayed to specific
players.

Plugin managed scoreboards are active so long as a reference is kept by a
plugin, or it has been registered as a player's active scoreboard. Objects
specific to a scoreboard remain active until unregistered (which remove a
reference to the owning scoreboard), but quickly fail if accessed
post-unregistration.

amaranth committed:
Update Bukkit for Minecraft 1.5.1

crast committed:
Don't cache metadata store disambiguations. Fixes BUKKIT-3841

The metadata system generates unique keys for metadata entries based on
the subject metadata is being applied to and the name of the metadata
being applied. It was assumed this would be an expensive operation so a
cache was put in place to ensure this was done as little as possible.

In reality this cache only has a benefit when you have a hit rate above
~90% and is otherwise much slower. As the implementation of the cache is
a hashmap of hashmaps it also uses a significant amount of memory which
is not worth it even for the performance increase with a high hit rate.

This commit simply removes the cache which results in speedups for most
cases and large memory savings.

Xephi committed:
Add Dropper BlockState.  Adds BUKKIT-3750

MisterVector committed:
Add method to get the source of a TNTPrimed.  Adds BUKKIT-3815

TriPhoenix committed:
Add InventoryMoveItemEvent. Adds BUKKIT-3765

This event is being called whenever a block or entity (e.g. hopper) tries to
move an item from one inventory to another inventory (one inventory may be
the hopper itself).

crast committed:
Prevent classloader leak in metadata system. Fixes BUKKIT-3854

Metadata values keep strong reference to plugins and they are not
cleared out when plugins are unloaded. This system adds weak reference
logic to allow these values to fall out of scope. In addition we get
some operations turning to O(1) "for free."

amaranth committed:
Map old wildcard recipe data to new 1.5 value. Fixes BUKKIT-3849

riking committed:
Add Thorns to DamageCause - Related to BUKKIT-3505

gmcferrin commented on src/main/java/org/bukkit/World.java:
```diff
@@ -1061,6 +1061,23 @@
+    
+    /**
+     * Broadcasts the specified message to every player in this world with the given permission
+     * 
```
Extra spaces.

gmcferrin commented on src/main/java/org/bukkit/World.java:
```diff
@@ -1061,6 +1061,23 @@
+    
+    /**
+     * Broadcasts the specified message to every player in this world
+     * 
```
Extra spaces.

SagaciousZed committed:
Pulling all pending Bukkit-JavaDoc changes

amaranth committed:
Add Beacon BlockState for hopper events. Fixes BUKKIT-3932

Commit:
Ignore all .DS_Store files, not just root folder

feildmaster committed:
Add Effect command. Adds BUKKIT-3763

amaranth committed:
Add method to update state without physics update. Addresses BUKKIT-3939

SagaciousZed committed:
Pulling all pending Bukkit-JavaDoc changes

EvilSeph committed:
Updated version to 1.5.1-R0.1 in pom.xml for BETA.

EvilSeph committed:
Updated version to 1.5.1-R0.2-SNAPSHOT for development towards next release.

Wolvereness committed:
Use utility method for team-join display. Fixes BUKKIT-3997

The method to make a string from a collection of strings already exists
and should be used when adding multiple players to a team.

Wolvereness committed:
Consider arguments to team leave properly. Fixes BUKKIT-3994

Two checks to argument length were changed to properly consider if the
sender is a player instead of an off-by-one logical error.

Acrobot committed:
Fix off-by-one error in DyeColor. Fixes BUKKIT-3938

Wolvereness committed:
Consider first player name in leave command. Fixes BUKKIT-4051

amaranth committed:
Correct 1.5 material data. Fixes BUKKIT-4004, BUKKIT-3785

Wolvereness committed:
Pulling all pending Bukkit-JavaDoc changes

EvilSeph committed:
Updated version to 1.5.1-R0.2 in pom.xml for BETA.

EvilSeph committed:
Updated version to 1.5.1-R0.3-SNAPSHOT for development towards next release.

riking committed:
Improve events for new inventory features. Adds BUKKIT-3859

This commit brings the InventoryClickEvent up to date with the new Minecraft
changes in 1.5.

InventoryDragEvent (thanks to @YLivay for his PR) is added to represent the
new "dragging" or "painting" functionality, where if you hold an itemstack and
click-drag over several slots, the items will be split evenly (left click) or
1 each (right click).

The ClickType enum is used to represent what the client did to trigger the
event.

The InventoryAction enum is reserved for future expansion, but will be used to
indicate the approximate result of the action.

Additionally, handling of creative inventory editing is improved with the new
InventoryCreativeEvent, and handling of numberkey presses is also improved
within InventoryClickEvent and CraftItemEvent.

Also, cancelling a creative click now displays properly on the client.

Adresses BUKKIT-3692, BUKKIT-4035, BUKKIT-3859 (new 1.5 events),
BUKKIT-2659, BUKKIT-3043, BUKKIT-2659, and BUKKIT-2897 (creative click events).

amaranth committed:
Update Bukkit for Minecraft 1.5.2

EvilSeph committed:
Updated version to 1.5.2-R0.1 in pom.xml for BETA.

EvilSeph committed:
Updated version to 1.5.2-R0.2-SNAPSHOT for development towards next release.

mbax committed:
Consider full team display name. Fixes BUKKIT-4186

Through a miscalculation, team display names were being created via
command ignoring the first word in the submitted display names.

Wolvereness committed:
Pulling all pending Bukkit-JavaDoc changes

Commit:
Fixed all spacing issues

Commit:
Merge branch 'master' of https://github.com/kezz101/Bukkit.git

Commit:
Fixed spacing issues

kezz commented:
I'm going to close this and reopen it as my git messed up.
